frontend/javascripts/components/color_picker.tsx

```diff
@@ -47,19 +47,26 @@ export function ChangeColorMenuItemContent({
 }: {
   title: string;
   isDisabled: boolean;
-  onSetColor: (rgb: Vector3) => void;
+  onSetColor: (rgb: Vector3, createsNewUndoState: boolean) => void;
   rgb: Vector3;
   hidePickerIcon?: boolean;
 }) {
+  const isFirstColorChange = useRef(true);
   const color = Utils.rgbToHex(Utils.map3((value) => value * 255, rgb));
   const onChangeColor = (colorStr: string) => {
     if (isDisabled) {
       return;
     }
     const colorRgb = Utils.hexToRgb(colorStr);
     const newColor = Utils.map3((component) => component / 255, colorRgb);
-    onSetColor(newColor);
+
+    // Only create a new undo state on the first color change event.
+    // All following color change events should mutate the most recent undo
+    // state so that the undo stack is not filled on each mouse movement.
```

Comment:
This works because the color picker is unmounted when it is closed (and remounted when it's opened again), right?

Reply:
Exactly! At first, I wanted to hook into the visibility-has-changed prop of the antd popover, but it turned out that wasn't even triggered, since the entire popover was simply unmounted when the menu closed. So, it's definitely unmounted :)